src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1153,7 +1153,17 @@ private HttpResponseHeaders CreateResponseHeaders(bool appCompleted)
         {
             if (!CanIncludeResponseContentLengthHeader())
             {
-                RejectInvalidHeaderForNonBodyResponse(appCompleted, HeaderNames.ContentLength);
+                if (responseHeaders.ContentLength.Value == 0)
+                {
+                    // If the response shouldn't include a Content-Length but it's 0
+                    // we'll just get rid of it without throwing an error, since it
+                    // is semantically equivalent to not having a Content-Length.
+                    responseHeaders.Remove(HeaderNames.ContentLength, out _);
```

Comment:
Why not `responseHeaders.ContentLength = null`? Or should both be here?

Reply:
They are kept in sync under the covers, e.g.: https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs#L13715-13723

I've changed it to `responseHeaders.ContentLength = null` since that's a bit cleaner.